client/cmd/dexc-desktop/pkg/Build-Windows.md

```diff
@@ -0,0 +1,36 @@
+## Windows build
+
+### Setting up the build environment
+
+ Download and run setup script:
+
+ ```batch
+ curl -O https://raw.githubusercontent.com/decred/dcrdex/windows-build/client/cmd/dexc-desktop/pkg/setup-windows.cmd
+ setup-windows.cmd <branch> <repoUrl>
+ ```
+
+ Both `branch` and `repoUrl` are optional.  If not specified, it will clone the `master` branch in the default [dcrdex repository](https://github.com/decred/dcrdex).
+
+ This will download and install `git` and `PowerShell`, clone the repo and install the build toolchain and required SDKs for the build.  Once completed, open a new command prompt in order to effectuate `PATH` and other environment variables configured by the setup script.
+
+ ### Running the build
+
+ ```batch
+ cd dcrdex\client\cmd\dexc-desktop
+ ```
+
+Build the Windows binary: 
+
+```batch
+pkg\build-windows.cmd
```

Comment:
At this step seeing:
```
decred.org/dcrdex/client/asset/firo
decred.org/dcrdex/client/asset/bch
decred.org/dcrdex/client/asset/ltc
decred.org/dcrdex/client/asset/zec
# github.com/webview/webview
In file included from webview.h:1056,
                 from webview.cc:1:
C:\Users\joe\git\dcrdex\client\cmd\dexc-desktop\build\libs\webview2\build\native\include/WebView2.h:1574:10: fatal error: EventToken.h: No such file or directory
 1574 | #include "EventToken.h"
      |          ^~~~~~~~~~~~~~
compilation terminated.
github.com/ethereum/go-ethereum/crypto
github.com/ethereum/go-ethereum/crypto/ecies
```

And the build fails after a bit.

Reply:
Looks like maybe gcc isn't on the PATH - maybe it's because the setup wasn't able to set it because it wasn't running as admin?  Did you open a new command prompt after the installation?

Reply:
> Did you open a new command prompt after the installation?

Yes, and it was admin. I remember seeing errors like this before... Do you want me to troubleshoot or do you have an idea? It's windows 10, I can provide more details.

Reply:
Did you restart `setup-windows.cmd` after the `PATH` permission issue above and it ran without error?  It really does seem like it's related to gcc (the MSYS2 subsystem) not being in place.  I can't replicate this locally, if you could provide details I'd appreciate that.

Reply:
> Did you restart setup-windows.cmd after the PATH permission issue above and it ran without error?

Yes.

Reply:
What Windows version/build are you running?

Reply:
Could you please try running the `go build -v ...` in `client/cmd/dexc-desktop` and see if you get the same error?

Reply:
> What Windows version/build are you running?

Version	10.0.19045 Build 19045

`go build -v ...` works.

Reply:
I see that header file in that location. I guess a problem with the enviroment.

Reply:
Trying with mingw64:
```
joe@DESKTOP-8Q0L2L5 MINGW64 /c/users/joe/git/dcrdex/client/cmd/dexc-desktop
# pkg/build-windows.cmd
Fetching and extracting the latest WebView2 SDK...
tar: This does not look like a tar archive
tar: Skipping to next header
tar: Exiting with failure status due to previous errors
Fetching and extracting WebView2 failed! Do you have tar and curl?

joe@DESKTOP-8Q0L2L5 MINGW64 /c/users/joe/git/dcrdex/client/cmd/dexc-desktop
# which tar
/usr/bin/tar

joe@DESKTOP-8Q0L2L5 MINGW64 /c/users/joe/git/dcrdex/client/cmd/dexc-desktop
# which curl
/usr/bin/curl
```

Reply:
I guess because the directory already exists. Gits the header error also if I comment out the tar.

Reply:
I commented out the tar, since the files are already there. Unsure why that fails. I was able to build after installing https://packages.msys2.org/package/mingw-w64-x86_64-headers-git and using mingw64.

At the end getting this message:
```
github.com/cockroachdb/pebble/sstable
github.com/cockroachdb/pebble
decred.org/dcrdex/client/cmd/dexc-desktop
Build completed
        1 file(s) copied.
Signing C:\Users\joe\git\dcrdex\client\cmd\dexc-desktop\build\windows\dexc-desktop.exe
Usage: sign-windows.cmd <file>
The WebView2Loader.dll file should be included with dexc-desktop.exe
```

Reply:
> I commented out the tar, since the files are already there. Unsure why that fails. 

The WebView NuGet package is compressed with BSD tar, MINGW has GNU tar:

https://github.com/chappjc/dcrdex/blob/linux-icon/client/cmd/dexc-desktop/build.cmd

> Build completed
>         1 file(s) copied.

For some reason `go` doesn't set `%ERRORLEVEL%`  properly, the script can't catch a failed build.

Reply:
> In file included from webview.h:1056,
>                  from webview.cc:1:
> C:\Users\joe\git\dcrdex\client\cmd\dexc-desktop\build\libs\webview2\build\native\include/WebView2.h:1574:10: fatal 
> error: EventToken.h: No such file or directory

This indicates that it wasn't able to extract the `Microsoft.Web.WebView2` package, though the script should error out.  If you run `tar` from a cmd prompt, which variant runs?

```
C:\Users\micro>tar --version
bsdtar 3.5.2 - libarchive 3.5.2 zlib/1.2.5.f-ipp
```

```
user@DESKTOP-FUKP40M UCRT64 ~
$ tar --version
tar (GNU tar) 1.35
```

Reply:
The versions of tar are the same as you. bsd for powershell and gnu for mingw.

Which cli should I be using?

How do I get that header into my powershell path?

Reply:
> Which cli should I be using?
> 
The command prompt.

> How do I get that header into my powershell path?

`pkg\build-windows.cmd` downloads and extracts it in the build directory, no need to do anything manually.  The script relies on BSD tar being on the `PATH` first (this is the default on Windows), so maybe if you have mingw tools earlier on the PATH then that might fail.

Do you have a virgin Win10 install you could try this on?  I made a few tweaks to the install script to make it more robust, it should really set up the environment properly.

Reply:
the bsd tar is on the path i think:
```
PS C:\Users\joe\git\dcrdex\client\cmd\dexc-desktop> tar --version
bsdtar 3.5.2 - libarchive 3.5.2 zlib/1.2.5.f-ipp
```

Reply:
does any of this look weird?
```
echo %PATH%
C:\Python310\Scripts\;C:\Python310\;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Windows\System32\OpenSSH\;C:\Program Files (x86)\NVIDIA Corporation\PhysX\Common;C:\ProgramData\chocolatey\bin;C:\TDM-GCC-64\bin;C:\Program Files\dotnet\;C:\Program Files (x86)\Windows Kits\10\Windows Performance Toolkit\;C:\Program Files\PowerShell\7\;C:\Program Files\Go\bin;C:\msys64\ucrt64\bin;C:\Program Files (x86)\nodejs\;C:\Program Files\Git\cmd;C:\msys64\ucrt64\bin;C:\Users\joe\AppData\Local\Microsoft\WindowsApps;C:\Users\joe\.dotnet\tools;C:\Users\joe\go\bin;C:\Users\joe\AppData\Roaming\npm
```

Reply:
Including the headers at `$MSysInstallPath\ucrt64\include` just leads to more errors. I've updated all I can think of. Unsure how to make it work.

Reply:
`C:\ProgramData\chocolatey\bin;C:\TDM-GCC-64\bin;` might cause issues depending on what's installed via choco.  I'm not sure if tdm-gcc will work, chappjc settled on mingw.

TBH the idea behind the automated setup script is that it's run on a clean install to ensure the integrity of the system when building release binaries so it's not super robust in terms of finding stuff that are already on the system.

Reply:
> Including the headers at `$MSysInstallPath\ucrt64\include` just leads to more errors. I've updated all I can think of. Unsure how to make it work.

Do you see any odd path in `gcc -v --version`?

Was msys installed on the system when you ran the setup script?

Reply:
> Was msys installed on the system when you ran the setup script?

Yes it already was. Should I try uninstalling first? Ill look at the gcc version soon and post back.

Reply:
Yes please, that's likely to solve the issue.

Reply:
<details closed>
<summary>gcc version</summary>

```
PS C:\Users\joe\git\dcrdex\client\cmd\dexc-desktop> gcc -v --version
‘g?Y???Y spec ?d?g?p?μ?A?￠?U?・?B
COLLECT_AS_OPTIONS='--version'
COLLECT_GCC=C:\TDM-GCC-64\bin\gcc.exe
COLLECT_LTO_WRAPPER=C:/TDM-GCC-64/bin/../libexec/gcc/x86_64-w64-mingw32/10.3.0/lto-wrapper.exe
gcc.exe (tdm64-1) 10.3.0
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


?^?[?Q?b?g: x86_64-w64-mingw32
configure ?Y’e: ../../../src/gcc-git-10.3.0/configure --build=x86_64-w64-mingw32 --enable-targets=all --enable-languages=ada,c,c++,fortran,jit,lto,objc,obj-c++ --enable-libgomp --enable-lto --enable-graphite --enable-cxx-flags=-DWINPTHREAD_STATIC --disable-build-with-cxx --disable-build-poststage1-with-cxx --enable-libstdcxx-debug --enable-threads=posix --enable-version-specific-runtime-libs --enable-fully-dynamic-string --enable-libstdcxx-filesystem-ts=yes --disable-libstdcxx-pch --enable-libstdcxx-threads --enable-libstdcxx-time=yes --enable-mingw-wildcard --with-gnu-ld --disable-werror --enable-nls --disable-win32-registry --enable-large-address-aware --disable-rpath --disable-symvers --prefix=/mingw64tdm --with-local-prefix=/mingw64tdm --with-pkgversion=tdm64-1 --with-bugurl=https://github.com/jmeubank/tdm-gcc/issues
?X???b?h???f??: posix
?T?|?[?g?3?e?A?￠?e LTO ?3?k?A???S???Y??: zlib zstd
gcc ?o?[?W?‡?“ 10.3.0 (tdm64-1)
COLLECT_GCC_OPTIONS='-v' '--version' '-mtune=generic' '-march=x86-64'
 C:/TDM-GCC-64/bin/../libexec/gcc/x86_64-w64-mingw32/10.3.0/cc1.exe -quiet -v -iprefix C:/TDM-GCC-64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/ -D_REENTRANT help-dummy -quiet -dumpbase help-dummy -mtune=generic -march=x86-64 -auxbase help-dummy -version --version -o C:\Users\joe\AppData\Local\Temp\ccv2sGTH.s
GNU C17 (tdm64-1) version 10.3.0 (x86_64-w64-mingw32)
        compiled by GNU C version 10.3.0, GMP version 6.2.1, MPFR version 4.1.0, MPC version 1.2.1, isl version isl-0.23-GMP

GGC heuristics: --param ggc-min-expand=100 --param ggc-min-heapsize=131072
COLLECT_GCC_OPTIONS='-v' '--version' '-mtune=generic' '-march=x86-64'
 C:/TDM-GCC-64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../x86_64-w64-mingw32/bin/as.exe -v --version -o C:\Users\joe\AppData\Local\Temp\cc03t0Bh.o C:\Users\joe\AppData\Local\Temp\ccv2sGTH.s
GNU assembler version 2.36.1 (x86_64-w64-mingw32) using BFD version (GNU Binutils) 2.36.1
GNU assembler (GNU Binutils) 2.36.1
Copyright (C) 2021 Free Software Foundation, Inc.
This program is free software; you may redistribute it under the terms of
the GNU General Public License version 3 or later.
This program has absolutely no warranty.
This assembler was configured for a target of `x86_64-w64-mingw32'.
COMPILER_PATH=C:/TDM-GCC-64/bin/../libexec/gcc/x86_64-w64-mingw32/10.3.0/;C:/TDM-GCC-64/bin/../libexec/gcc/;C:/TDM-GCC-64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../x86_64-w64-mingw32/bin/
LIBRARY_PATH=C:/TDM-GCC-64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/;C:/TDM-GCC-64/bin/../lib/gcc/;C:/TDM-GCC-64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../x86_64-w64-mingw32/lib/../lib/;C:/TDM-GCC-64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../lib/;C:/TDM-GCC-64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../x86_64-w64-mingw32/lib/;C:/TDM-GCC-64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../
COLLECT_GCC_OPTIONS='-v' '--version' '-mtune=generic' '-march=x86-64'
 C:/TDM-GCC-64/bin/../libexec/gcc/x86_64-w64-mingw32/10.3.0/collect2.exe -plugin C:/TDM-GCC-64/bin/../libexec/gcc/x86_64-w64-mingw32/10.3.0/liblto_plugin-0.dll -plugin-opt=C:/TDM-GCC-64/bin/../libexec/gcc/x86_64-w64-mingw32/10.3.0/lto-wrapper.exe -plugin-opt=-fresolution=C:\Users\joe\AppData\Local\Temp\ccdKaDRZ.res -plugin-opt=-pass-through=-lmingw32 -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lpthread -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lkernel32 -plugin-opt=-pass-through=-lmoldname -plugin-opt=-pass-through=-lmingwex -plugin-opt=-pass-through=-lmsvcrt -plugin-opt=-pass-through=-lkernel32 -plugin-opt=-pass-through=-lpthread -plugin-opt=-pass-through=-ladvapi32 -plugin-opt=-pass-through=-lshell32 -plugin-opt=-pass-through=-luser32 -plugin-opt=-pass-through=-lkernel32 -plugin-opt=-pass-through=-lmingw32 -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lpthread -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lkernel32 -plugin-opt=-pass-through=-lmoldname -plugin-opt=-pass-through=-lmingwex -plugin-opt=-pass-through=-lmsvcrt -plugin-opt=-pass-through=-lkernel32 -m i386pep --exclude-libs=libpthread.a --undefined=__xl_f -Bdynamic --version C:/TDM-GCC-64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../x86_64-w64-mingw32/lib/../lib/crt2.o C:/TDM-GCC-64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/crtbegin.o -LC:/TDM-GCC-64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0 -LC:/TDM-GCC-64/bin/../lib/gcc -LC:/TDM-GCC-64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../x86_64-w64-mingw32/lib/../lib -LC:/TDM-GCC-64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../lib -LC:/TDM-GCC-64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../x86_64-w64-mingw32/lib -LC:/TDM-GCC-64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../.. C:\Users\joe\AppData\Local\Temp\cc03t0Bh.o -lmingw32 -lgcc -lpthread -lgcc -lkernel32 -lmoldname -lmingwex -lmsvcrt -lkernel32 -lpthread -ladvapi32 -lshell32 -luser32 -lkernel32 -lmingw32 -lgcc -lpthread -lgcc -lkernel32 -lmoldname -lmingwex -lmsvcrt -lkernel32 C:/TDM-GCC-64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../x86_64-w64-mingw32/lib/../lib/default-manifest.o C:/TDM-GCC-64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/crtend.o
collect2 version 10.3.0
C:/TDM-GCC-64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe -plugin C:/TDM-GCC-64/bin/../libexec/gcc/x86_64-w64-mingw32/10.3.0/liblto_plugin-0.dll -plugin-opt=C:/TDM-GCC-64/bin/../libexec/gcc/x86_64-w64-mingw32/10.3.0/lto-wrapper.exe -plugin-opt=-fresolution=C:\Users\joe\AppData\Local\Temp\ccdKaDRZ.res -plugin-opt=-pass-through=-lmingw32 -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lpthread -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lkernel32 -plugin-opt=-pass-through=-lmoldname -plugin-opt=-pass-through=-lmingwex -plugin-opt=-pass-through=-lmsvcrt -plugin-opt=-pass-through=-lkernel32 -plugin-opt=-pass-through=-lpthread -plugin-opt=-pass-through=-ladvapi32 -plugin-opt=-pass-through=-lshell32 -plugin-opt=-pass-through=-luser32 -plugin-opt=-pass-through=-lkernel32 -plugin-opt=-pass-through=-lmingw32 -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lpthread -plugin-opt=-pass-through=-lgcc -plugin-opt=-pass-through=-lkernel32 -plugin-opt=-pass-through=-lmoldname -plugin-opt=-pass-through=-lmingwex -plugin-opt=-pass-through=-lmsvcrt -plugin-opt=-pass-through=-lkernel32 -m i386pep --exclude-libs=libpthread.a --undefined=__xl_f -Bdynamic --version C:/TDM-GCC-64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../x86_64-w64-mingw32/lib/../lib/crt2.o C:/TDM-GCC-64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/crtbegin.o -LC:/TDM-GCC-64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0 -LC:/TDM-GCC-64/bin/../lib/gcc -LC:/TDM-GCC-64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../x86_64-w64-mingw32/lib/../lib -LC:/TDM-GCC-64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../lib -LC:/TDM-GCC-64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../x86_64-w64-mingw32/lib -LC:/TDM-GCC-64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../.. C:\Users\joe\AppData\Local\Temp\cc03t0Bh.o -lmingw32 -lgcc -lpthread -lgcc -lkernel32 -lmoldname -lmingwex -lmsvcrt -lkernel32 -lpthread -ladvapi32 -lshell32 -luser32 -lkernel32 -lmingw32 -lgcc -lpthread -lgcc -lkernel32 -lmoldname -lmingwex -lmsvcrt -lkernel32 C:/TDM-GCC-64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../x86_64-w64-mingw32/lib/../lib/default-manifest.o C:/TDM-GCC-64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/crtend.o
GNU ld (GNU Binutils) 2.36.1
Copyright (C) 2021 Free Software Foundation, Inc.
This program is free software; you may redistribute it under the terms of
the GNU General Public License version 3 or (at your option) a later version.
This program has absolutely no warranty.
COLLECT_GCC_OPTIONS='-v' '--version' '-mtune=generic' '-march=x86-64'
</details>

Reply:
Tried deleting msys and restarting and running the scripts again but stops at the same place. `EventToken.h: No such file or directory`

Reply:
Is there another source for https://www.nuget.org/api/v2/package/Microsoft.Web.WebView2 ?

Reply:
> Tried deleting msys and restarting and running the scripts again but stops at the same place. `EventToken.h: No such file or directory`


```
C:\Users\user>dir /s c:\EventToken.h
...
 Directory of c:\msys64\ucrt64\include

12/04/2023  06:59 AM               882 eventtoken.h
               1 File(s)            882 bytes
```

Maybe something in  `PATH` lingering around?

Would you mind testing this in a fresh Windows install?  I'll launch a VPS and give you the RDP details.

Reply:
> Is there another source for https://www.nuget.org/api/v2/package/Microsoft.Web.WebView2 ?

This is the official source.

https://learn.microsoft.com/en-us/microsoft-edge/webview2/how-to/machine-setup#install-or-update-the-webview2-sdk

Reply:
> Would you mind testing this in a fresh Windows install? I'll launch a VPS and give you the RDP details.

If a user finds themselves in my situation, that doesn't really help though. Maybe someone else @buck54321 @martonp @itswisdomagain or whoever can try on their machine. Will end users be using this installer, or just whoever makes the packages that end users would install.

Reply:
> Maybe something in PATH lingering around?

I see `C:\msys64\ucrt64\bin` I guess for every time I used the script. I see eventtoken.h in the include folder.

<details closed>
<summary>path</summary>

```
$env:path -split ";"
C:\Python310\Scripts\
C:\Python310\
C:\Windows\system32
C:\Windows
C:\Windows\System32\Wbem
C:\Windows\System32\WindowsPowerShell\v1.0\
C:\Windows\System32\OpenSSH\
C:\Program Files (x86)\NVIDIA Corporation\PhysX\Common
C:\ProgramData\chocolatey\bin
C:\TDM-GCC-64\bin
C:\Program Files\dotnet\
C:\Program Files\PowerShell\7\
C:\Program Files\Go\bin
C:\msys64\ucrt64\bin
C:\Program Files (x86)\nodejs\
C:\msys64\ucrt64\bin
C:\msys64\ucrt64\bin
C:\Program Files (x86)\Windows Kits\10\Windows Performance Toolkit\
C:\msys64\ucrt64\bin
C:\Program Files\Git\cmd
C:\msys64\ucrt64\bin
C:\Users\joe\AppData\Local\Microsoft\WindowsApps
C:\Users\joe\.dotnet\tools
C:\Users\joe\go\bin
C:\Users\joe\AppData\Roaming\npm
```
</details>

Reply:
Please see my newer comment I guess it's solved for me. Just need to add to the includes. I tried this before but couldn't figure out how to write it last time.

Reply:
> If a user finds themselves in my situation, that doesn't really help though. Maybe someone else @buck54321 @martonp @itswisdomagain or whoever can try on their machine. Will end users be using this installer, or just whoever makes the packages that end users would install.

It's not meant for the end user, only for internal use to build the binary and the installer.  With Windows being as vulnerable as it is, it's important to have a VM only for this single purpose (to make sure no malware can creep in) so there really shouldn't be anything installed on the system when the setup script and the build are run.

Reply:
Oh I see.